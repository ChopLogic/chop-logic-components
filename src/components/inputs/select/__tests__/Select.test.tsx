import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import ChopLogicForm from 'components/containers/form/Form';

import ChopLogicSelect, { SelectValue } from '../Select';

describe('ChopLogicSelect', () => {
  const SELECT_LANGUAGES: SelectValue[] = [
    { id: '57b5a856-bc4b-4c2f-8295-9f44434fafc7', label: 'English' },
    { id: 'b4b08a09-e287-400b-ba65-afd6ae1f6fd7', label: 'Russian' },
    { id: '641fc722-af5a-4426-8f8a-1566cc492b91', label: 'German' },
  ];

  const testProps = {
    id: 'language-select-id',
    name: 'language',
    label: 'Select your language',
    options: SELECT_LANGUAGES,
    onChange: vi.fn(),
    placeholder: 'Not selected',
  };

  it('should render correctly', () => {
    const { asFragment } = render(<ChopLogicSelect {...testProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should display the label', () => {
    render(<ChopLogicSelect {...testProps} />);
    expect(screen.getByLabelText(testProps.label)).toBeInTheDocument();
  });

  it('should have accessible roles', async () => {
    render(<ChopLogicSelect {...testProps} />);
    expect(screen.getByRole('combobox')).toBeInTheDocument();
    await userEvent.click(screen.getByRole('combobox'));
    expect(screen.getByRole('listbox')).toBeInTheDocument();
  });

  it('could be disabled', () => {
    render(<ChopLogicSelect {...testProps} disabled />);
    expect(screen.getByRole('combobox')).toBeDisabled();
  });

  it('could be required', () => {
    render(<ChopLogicSelect {...testProps} required />);
    expect(screen.getByRole('combobox')).toHaveAttribute('aria-required', 'true');
  });

  it('should display all values as options', async () => {
    render(<ChopLogicSelect {...testProps} />);
    await userEvent.click(screen.getByRole('combobox'));
    const options = screen.getAllByRole('option');
    expect(options).toHaveLength(SELECT_LANGUAGES.length);
    options.forEach((option, index) => {
      expect(option).toHaveTextContent(SELECT_LANGUAGES[index].label);
    });
  });

  it('should call onSelectChange handler', async () => {
    render(<ChopLogicSelect {...testProps} />);
    const combobox = screen.getByRole('combobox');
    await userEvent.click(combobox);
    const option = screen.getByText(SELECT_LANGUAGES[0].label);
    await userEvent.click(option);
    expect(testProps.onChange).toHaveBeenCalledOnce();
  });

  it('should allow the user to select an option', async () => {
    render(<ChopLogicSelect {...testProps} />);
    const combobox = screen.getByRole('combobox');
    expect(combobox).toHaveValue('');
    await userEvent.click(combobox);
    const option = screen.getByText(SELECT_LANGUAGES[1].label);
    await userEvent.click(option);
    expect(combobox).toHaveValue(SELECT_LANGUAGES[1].id);
  });

  it('should allow the user to deselect the same option', async () => {
    render(<ChopLogicSelect {...testProps} />);
    const combobox = screen.getByRole('combobox');
    expect(combobox).toHaveValue('');
    await userEvent.click(combobox);
    const option = screen.getByText(SELECT_LANGUAGES[1].label);
    await userEvent.click(option);
    expect(combobox).toHaveValue(SELECT_LANGUAGES[1].id);
    await userEvent.click(option);
    expect(combobox).not.toHaveValue(SELECT_LANGUAGES[1].id);
  });

  it('should move the focus correctly on Tab press', async () => {
    render(<ChopLogicSelect {...testProps} />);
    const combobox = screen.getByRole('combobox');
    await userEvent.click(combobox);

    const options = screen.getAllByRole('option');
    options[0].focus();
    await userEvent.tab();
    expect(options[1]).toHaveFocus();
    await userEvent.tab();
    expect(options[2]).toHaveFocus();
    await userEvent.tab();
    expect(options[0]).toHaveFocus();
  });

  it('should move focus to the next option by pressing ArrowDown button', async () => {
    render(<ChopLogicSelect {...testProps} />);
    await userEvent.click(screen.getByRole('combobox'));
    const options = screen.getAllByRole('option');
    options[0].focus();
    await userEvent.keyboard('[ArrowDown]');
    expect(options[1]).toHaveFocus();
  });

  it('should move focus to the previous option by pressing ArrowUp button', async () => {
    render(<ChopLogicSelect {...testProps} />);
    await userEvent.click(screen.getByRole('combobox'));
    const options = screen.getAllByRole('option');
    options[1].focus();
    await userEvent.keyboard('[ArrowUp]');
    expect(options[0]).toHaveFocus();
  });

  it('should take an initial value from the form context', async () => {
    render(
      <ChopLogicForm initialValues={{ language: SELECT_LANGUAGES[2] }}>
        <ChopLogicSelect {...testProps} />
      </ChopLogicForm>,
    );
    const combobox = screen.getByRole('combobox');
    expect(combobox).toHaveValue(SELECT_LANGUAGES[2].id);
  });
});
