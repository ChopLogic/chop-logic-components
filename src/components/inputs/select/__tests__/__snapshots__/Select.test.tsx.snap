// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChopLogicSelect component > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="_wrapper_b2e794"
  >
    <label
      class="_label_c383e9 _label_b2e794"
      for="test-select-id_combobox"
    >
      <span>
        Select your language
      </span>
    </label>
    <button
      aria-controls="test-select-id_dropdown"
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-label="Select one of the options"
      aria-required="false"
      class="_combobox_b2e794"
      id="test-select-id_combobox"
      name="language"
      role="combobox"
      type="button"
    >
      <span
        class="_combobox_label_b2e794"
      >
        Not selected yet
      </span>
      <span
        aria-hidden="true"
        class="_icon_b2e794 chop-icon__caret-down"
      />
    </button>
    <ul
      class="_dropdown_b2e794"
      id="test-select-id_dropdown"
      role="listbox"
      tabindex="-1"
    >
      <li
        aria-selected="false"
        class="_option_b2e794"
        id="57b5a856-bc4b-4c2f-8295-9f44434fafc7"
        role="option"
        tabindex="0"
      >
        <span>
          English
        </span>
      </li>
      <li
        aria-selected="false"
        class="_option_b2e794"
        id="b4b08a09-e287-400b-ba65-afd6ae1f6fd7"
        role="option"
        tabindex="0"
      >
        <span>
          Russian
        </span>
      </li>
      <li
        aria-selected="false"
        class="_option_b2e794"
        id="641fc722-af5a-4426-8f8a-1566cc492b91"
        role="option"
        tabindex="0"
      >
        <span>
          German
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
